segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1432 @@
+                checkSegmentExists(streamSegmentName, segmentMetadata);
+                segmentMetadata.checkInvariants();
+                // This segment was created by an older segment store. Need to start a fresh new chunk.
+                if (segmentMetadata.getOwnerEpoch() < this.epoch) {
+                    log.debug("{} openWrite - Segment needs ownership change - segment={}.", logPrefix, segmentMetadata.getName());
+                    claimOwnership(txn, segmentMetadata);
+                }
+                // If created by newer instance then abort.
+                if (segmentMetadata.getOwnerEpoch() > this.epoch) {
+                    throw new StorageNotPrimaryException(streamSegmentName);
+                }
+
+                // This instance is the owner, return a handle.
+                val retValue = SegmentStorageHandle.writeHandle(streamSegmentName);
+                LoggerHelpers.traceLeave(log, "openWrite", traceId, retValue);
+                return retValue;
+            } catch (StorageMetadataWritesFencedOutException ex) {
+                throw new StorageNotPrimaryException(streamSegmentName, ex);
+            }
+        });
+    }
+
+    /**
+     * Checks ownership and adjusts the length of the segment if required.
+     *
+     * @param txn             Active {@link MetadataTransaction}.
+     * @param segmentMetadata {@link SegmentMetadata} for the segment to change ownership for.
+     * @throws Exception In case of any errors.
+     */
+    private void claimOwnership(MetadataTransaction txn, SegmentMetadata segmentMetadata) throws Exception {
+        // Claim ownership.
+        // This is safe because the previous instance is definitely not an owner anymore. (even if this instance is no more owner)
+        segmentMetadata.setOwnerEpoch(this.epoch);
+        segmentMetadata.setOwnershipChanged(true);
+
+        // Get the last chunk
+        String lastChunkName = segmentMetadata.getLastChunk();
+        if (null != lastChunkName) {
+            ChunkMetadata lastChunk = (ChunkMetadata) txn.get(lastChunkName);
+            log.debug("{} claimOwnership - current last chunk - segment={}, last chunk={}, Length={}.",
+                    logPrefix,
+                    segmentMetadata.getName(),
+                    lastChunk.getName(),
+                    lastChunk.getLength());
+            ChunkInfo chunkInfo = chunkStorage.getInfo(lastChunkName);
+            Preconditions.checkState(chunkInfo != null);
+            Preconditions.checkState(lastChunk != null);
+            // Adjust its length;
+            if (chunkInfo.getLength() != lastChunk.getLength()) {
+                Preconditions.checkState(chunkInfo.getLength() > lastChunk.getLength());
+                // Whatever length you see right now is the final "sealed" length of the last chunk.
+                lastChunk.setLength((int) chunkInfo.getLength());
```

Comment:
this cast  looks very dangerous.
A chunk is expected to be longer than Integer.MAX_VALUE

Reply:
Good catch. That cast is not needed at all. (will fix)

Reply:
I am actually surprised that spotbugs didn't catch it.

Reply:
fixed